### ISSUE TITLE
[CAPI] lbfgs doesn't work on gpu.

### ISSUE THREAD

**Issue by Strasser-Pablo** [pytorch/pytorch#25854]
## 🐛 Bug

LBFGS doesn't work on loss function returning value on GPU.

Using the gpu give the following stack trace:

> terminate called after throwing an instance of 'c10::Error'
>   what():  expected device cuda:0 and dtype Float but got device cpu and dtype Float (compute_types at ../aten/src/ATen/native/TensorIterator.cpp:171)
> frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x6c (0x7f45544a64cc in /usr/local/lib/python3.6/dist-packages/torch/lib/libc10.so)
> frame #1: at::TensorIterator::compute_types() + 0xb09 (0x7f4557b123d9 in /usr/local/lib/python3.6/dist-packages/torch/lib/libtorch.so)
> frame #2: at::TensorIterator::Builder::build() + 0x55 (0x7f4557b170a5 in /usr/local/lib/python3.6/dist-packages/torch/lib/libtorch.so)
> frame #3: at::TensorIterator::binary_op(at::Tensor&, at::Tensor const&, at::Tensor const&) + 0x340 (0x7f4557b17a30 in /usr/local/lib/python3.6/dist-packages/torch/lib/libtorch.so)
> frame #4: at::native::mul(at::Tensor const&, at::Tensor const&) + 0xc7 (0x7f4557850e27 in /usr/local/lib/python3.6/dist-packages/torch/lib/libtorch.so)
> frame #5: <unknown function> + 0x57753a8 (0x7f4559e343a8 in /usr/local/lib/python3.6/dist-packages/torch/lib/libtorch.so)
> frame #6: torch::autograd::VariableType::mul(at::Tensor const&, at::Tensor const&) + 0x53f (0x7f455994520f in /usr/local/lib/python3.6/dist-packages/torch/lib/libtorch.so)
> frame #7: at::Tensor::mul(at::Tensor const&) const + 0x61 (0x4e90d1 in ./SmileGRUZincGenerate)
> frame #8: ./SmileGRUZincGenerate() [0x4e749e]
> frame #9: ./SmileGRUZincGenerate() [0x4e11ae]
> frame #10: ./SmileGRUZincGenerate() [0x420839]
> frame #11: __libc_start_main + 0xe7 (0x7f454e36bb97 in /lib/x86_64-linux-gnu/libc.so.6)
> frame #12: ./SmileGRUZincGenerate() [0x42089a]

For testing I changed line 118 of lbfgs.cpp from
`t = torch::tensor(options.learning_rate_, torch::kFloat32);`
to 
`t = torch::tensor(options.learning_rate_, at::kCUDA);`

And this solve the issue.

cc @yf225 @vincentqb
